clients/drcachesim/tracer/raw2trace.h

```diff
@@ -1727,6 +1729,70 @@ template <typename T> class trace_converter_t {
 #undef DR_CHECK
 };
 
+// We need to determine the memref_t record count for inserting a marker with
+// that count at the start of each chunk.
+class memref_counter_t : public reader_t {
+public:
+    bool
+    init() override
+    {
+        return true;
+    }
+    trace_entry_t *
+    read_next_entry() override
+    {
+        return nullptr;
+    };
+    bool
+    read_next_thread_entry(size_t thread_index, OUT trace_entry_t *entry,
+                           OUT bool *eof) override
+    {
+        return false;
+    }
+    std::string
+    get_stream_name() const override
+    {
+        return "";
+    }
+    int
+    entry_memref_count(const trace_entry_t *entry)
+    {
+        // Mirror file_reader_t::open_input_files().
+        // In particular, we need to skip TRACE_TYPE_HEADER and to pass the
+        // tid and tid to the reader before the 2 markers in front of them.
+        if (!saw_pid_) {
+            if (entry->type == TRACE_TYPE_HEADER)
+                return 0;
+            else if (entry->type == TRACE_TYPE_THREAD) {
+                list_.push_front(*entry);
+                return 0;
+            } else if (entry->type != TRACE_TYPE_PID) {
+                list_.push_back(*entry);
+                return 0;
+            }
+            saw_pid_ = true;
```

Comment:
It would be less confusing if saw_pid_ is set in the `entry->type != TRACE_TYPE_PID` else-if block above, and  the `saw_pid_` case below (L1785-) adds the currently entry to list_ and processes the whole list (like the current 1774-1783).  Can remove the current 1774-1783 then. Or am I missing some edge case?

Reply:
But then the common case (for millions of records: all but the 1st 5) would be to add one entry to an empty list, and then immediately remove that entry from the list.  That doesn't sound very efficient.